### PR TITLE
Test fsdp compile

### PR DESCRIPTION
**Pull request by y-sq**
- Added an initial benchmark for fsdp + fp8, benchmarks/bench_multi_gpu.py
  - Note that the benchmark has issues with "compile", please see the experiments doc for more context.
  - CUDA_VISIBLE_DEVICES=0,1 python benchmarks/bench_multi_gpu.py 2>&1 | tee logs

- Modified the existing test/test_fsdp.py to include compile.
  - The script can success with compile, but it's with a workaround. Please see the comment there for context.
  - Added relu layers after linear layers so that the numerics can be stable after several iterations and can be used to check accuracy results.

- In bench/profile_linear_float8.py, changed the TE FP8 receipt to be the same as PyTorch FP8.

- In sync_float8_amax_and_scale_history, added an option to combine all_reduce calls into one call.
  - Need to verify if this can bring performance gain.
  - Another option is to let each layer handle its own amax update and sync, which is the approach used by TE fp8.

### CODE REVIEW

**Comment by y-sq** [comment]
For the failed check, I verified "ufmt check ." locally which gave no error.

File: test/test_fsdp.py [hunk 1]
```diff
@@ -83,7 +86,7 @@ def fsdp_main(rank, world_size, args):
         rank
     )
     model.load_state_dict(torch.load(sd_in_fname))
-    model = FSDP(model)
+    model = FSDP(model, use_orig_params=COMPILE)
```
Comment (vkuzo) [inline test/test_fsdp.py:89]: just set use_orig_params to True, and add a comment that compile only supports this setting?

File: test/test_fsdp.py [hunk 2]
```diff
@@ -97,12 +100,21 @@ def fsdp_main(rank, world_size, args):
...
         y_local.sum().backward()
         sync_float8_amax_and_scale_history(model)
         optimizer.step()
+        return y_local
+
+    # TODO: Fix here. Running for one iteration and then compile is a workaround for compiling fsdp float8.
+    # A better way for compile is to FSDP(torch.compile(model)).
+    forward_backward()
+    if COMPILE:
+        forward_backward = torch.compile(forward_backward)
```
Comment (vkuzo) [inline test/test_fsdp.py:115]: are we sure this is supported?

how about something simpler such as

```
model = torch.compile(model)
sync_func = torch.compile(sync_float8_amax_and_scale_history)
for _ in range(N_ITER):
   ... loop as before ...
```

File: float8_experimental/float8_linear_utils.py [hunk 3]
```diff
@@ -117,8 +117,23 @@ def swap_linear_with_float8_linear(
...
+    fp8_layers = [
+        child for name, child in model.named_modules() if isinstance(child, fp8_classes)
+    ]
+
+    return fp8_layers
+
+
 def sync_float8_amax_and_scale_history(
-    model: torch.nn.Module, fp8_classes=None
+    model: torch.nn.Module, fp8_classes=None, fp8_layers=None, combine_reduction=False
```
Comment (vkuzo) [inline float8_experimental/float8_linear_utils.py:136]: thoughts on splitting these changes into a separate diff? something like
1. diff 1: add microbenchmark for FSDP
2. diffs 2..n: improvements to sync function, with before+after benchmark numbers

File: benchmarks/bench_multi_gpu.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+def run():
+    compile = True
+    base_dtype = torch.bfloat16
+    print(f"{base_dtype = }")
+    print(f"{compile = }")
+    # generate input data
+    ref_input = torch.randn(B, M, K).cuda().to(base_dtype)
+    # run fsdp model
+    WORLD_SIZE = torch.cuda.device_count()
```
Comment (vkuzo) [inline benchmarks/bench_multi_gpu.py:234]: maybe make this configurable? usually easier to debug issues and iterate with world_size == 2, and then verify the fix with world_size == torch.cuda.device_count()

File: benchmarks/bench_multi_gpu.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    print(f"{WORLD_SIZE = }")
+    args = (base_dtype, ref_input, compile)
+    mp.spawn(fsdp_main, args=(WORLD_SIZE, args), nprocs=WORLD_SIZE, join=True)
+
+
+if __name__ == "__main__":
```
Comment (vkuzo) [inline benchmarks/bench_multi_gpu.py:240]: does this need a shell script wrapper like the other distributed benchmarks in this file?  if not, can you add a comment with example way to call this?

File: float8_experimental/float8_linear_utils.py (hunk 3 above)
Comment (y-sq) [reply float8_experimental/float8_linear_utils.py:136]: Thank you for the comments. Let me split into small PRs.

File: test/test_fsdp.py (hunk 2 above)
Comment (y-sq) [reply test/test_fsdp.py:115]: This did work in fsdp.py test. But let me change it to the officially supported `FSDP(torch.compile(model))`.

**Comment by y-sq** [comment]
Will split and close this one.
